### PR TITLE
Dns over http2 5773 v12

### PR DESCRIPTION
**Pull request by catenacyber**
Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
https://redmine.openinfosecfoundation.org/issues/5773

Describe changes:
- analyze DNS over HTTP2

SV_BRANCH=https://github.com/OISF/suricata-verify/pull/1734

Draft to get feedback about approach...

https://github.com/OISF/suricata/pull/11242 with needed rebase

TODO : 
- Fix dns json schema and logging first (that is https://redmine.openinfosecfoundation.org/issues/6281 )
- Use config for dns logging...

Functionnaly, in terms of output :
- The flow will have `doh2` as app_proto (and `http2` as `app_proto_orig`)
- There are `doh2` events that have both http2 and dns fields. dns logging is done like alerts, not like dns events...
- DNS and HTTP and HTTP2 signatures work on DOH2
- Signatures can be DOH2 specific. These signatures can combine http, http2 and dns keywords.
- DNS Frames do not work on DoH2

Memory management
- a HTTP2 tx can own 2 DNS tx (one to client, one to server)
- a HTTP2 tx stores the streamed DNS response content until it is complete before parsing it (limiting to U16_MAX bytes)

API
- There is a new DOH2 app-layer protocol which resorts to HTTP2 for most of the things
- HTTP2 parsing discovering DOH2 changes the app_proto to DOH2, doh2 can be enabled or disabled in suricata.yaml config
- There are more alproto "comparison" functions
- Every DNS keyword needs to check the flow alproto to change the transaction
- Every DNS/HTTP keyword is automatically registered for DOH2 (hacking DetectAppLayerMpmRegister2...

### CODE REVIEW

**Comment by suricata-qa** [comment]
Information: QA ran without warnings. 

Pipeline 21055

File: rust/src/http2/http2.rs [hunk 1]
```diff
@@ -146,6 +149,12 @@ pub struct HTTP2Transaction {
     pub escaped: Vec<Vec<u8>>,
     pub req_line: Vec<u8>,
     pub resp_line: Vec<u8>,
+
+    is_doh_response: bool,
```
Comment (victorjulien) [inline rust/src/http2/http2.rs]: should this all be wrapped in an Option? It seems like a large expansion of the http/2 tx

File: rust/src/http2/logger.rs [hunk 2]
```diff
@@ -281,6 +282,35 @@ fn log_http2(tx: &HTTP2Transaction, js: &mut JsonBuilder) -> Result<bool, JsonEr
...
 
+    if tx.dns_request_tx.is_some() || tx.dns_response_tx.is_some() {
+        js.open_object("dns")?;
+        if let Some(dtx) = &tx.dns_request_tx {
+            let mark = js.get_mark();
+            let mut has_dns_query = false;
+            js.open_array("query")?;
+            for i in 0..0xFFFF {
+                let mut jsa = JsonBuilder::try_new_object()?;
+                if !SCDnsLogJsonQuery(dtx, i, 0xFFFFFFFFFFFFFFFF, &mut jsa) {
```
Comment (victorjulien) [inline rust/src/http2/logger.rs:293]: what are these 0xFFFFFFFFFFFFFFFF uses, looks not very nice

File: src/detect-dns-answer-name.c [hunk 3]
```diff
@@ -56,6 +56,12 @@ static InspectionBuffer *GetBuffer(DetectEngineThreadCtx *det_ctx,
         return buffer;
     }
 
+    if (f->alproto == ALPROTO_DOH2) {
```
Comment (victorjulien) [inline src/detect-dns-answer-name.c:59]: can we hide this behind this a single (inline?) call in all places?

**Comment by victorjulien** [comment]
looks good overall. Some comments inline.

File: rust/src/http2/logger.rs (hunk 2 above)
Comment (catenacyber) [reply rust/src/http2/logger.rs:293]: Indeed not very nice, this PR is still a draft waiting for the dns log overhaul in https://redmine.openinfosecfoundation.org/issues/6281 to do the same thing...

File: rust/src/http2/http2.rs (hunk 1 above)
Comment (catenacyber) [reply rust/src/http2/http2.rs]: Looks like a good idea, will try to improve on it

File: src/detect-dns-answer-name.c (hunk 3 above)
Comment (catenacyber) [reply src/detect-dns-answer-name.c:59]: Will try to make it nicer...

**Pull request closed without merging**